ansible_runner/receptor_plugin.py

```diff
@@ -0,0 +1,164 @@
+import logging
+import json
+import os
+import time
+import io
+import tarfile
+import tempfile
+import uuid
+import asyncio
+
+try:
+    import receptor
+    from receptor.config import ReceptorConfig
+    from receptor.controller import Controller
+    receptor_import = True
+except ImportError:
+    receptor_import = False
+
+
+from ansible_runner import run
+
+
+logger = logging.getLogger(__name__)
+
+
+class Receptor_Runner_Result:
+
+    def __init__(self):
+        self.rc = None
+
+
+class UUIDEncoder(json.JSONEncoder):
+    def default(self, obj):
+        if isinstance(obj, uuid.UUID):
+            return obj.hex
+        return json.JSONEncoder.default(self, obj)
+
+
+def run_via_receptor(receptor_node, receptor_peer, receptor_node_id, run_options):
+
+    async def read_responses():
+        while True:
+            message = await controller.recv()
+            if message.header.get("eof", False):
+                break
+            elif message.payload:
+                content = json.loads(message.payload.readall())
+                c_header = content[0]
+                c_type = c_header['type']
+                if c_type == 'event':
+                    data = content[1]
+                    if 'event_handler' in run_options and run_options['event_handler']:
+                        run_options['event_handler'](data)
+                    if 'stdout' in data and data['stdout']:
+                        print(data['stdout'])
+                elif c_type == 'status':
+                    data = content[1]
+                    if 'status_handler' in run_options and run_options['status_handler']:
+                        run_options['status_handler'](data, None)
+                elif c_type == 'error':
+                    print("Error from remote:", content[1])
+
+    async def run_func():
+        if receptor_node_id != receptor_node:
+            add_peer_task = controller.add_peer(receptor_peer)
+            start_wait = time.time()
+            while True:
+                if add_peer_task and add_peer_task.done() and not add_peer_task.result():
+                    return False
+                if controller.receptor.router.node_is_known(receptor_node):
+                    break
+                if time.time() - start_wait > 5:
+                    if not add_peer_task.done():
+                        add_peer_task.cancel()
+                    return False
+                await asyncio.sleep(0.1)
+        await controller.send(payload=tmpf.name, recipient=receptor_node, directive='ansible_runner:execute')
+        await controller.loop.create_task(read_responses())
+
+    remote_options = {key: value for key, value in run_options.items() if key in (
+        'forks',
+        'host_pattern',
+        'ident',
+        'ignore_logging',
+        'inventory',
+        'limit',
+        'module',
+        'module_args',
+        'omit_event_data',
+        'only_failed_event_data',
+        'playbook',
+        'verbosity',
```

Comment:
Is this list hand-curated from somewhere, and is there a reason we're whitelisting things?  It's kind of gross that `ansible_runner.run` is just a giant docstring that takes blanket `**kwargs`, but is there some underlying list of these define in code somewhere that we could glean using `inspect.getargspec`?

Reply:
Curated is too strong a word.  It's the parameters that I thought made sense to run remotely.  I specifically want to avoid trying to pass along handlers that can't be called from the remote side, or anything that isn't json serializable.  Will look into finding a better way to do this.

Reply:
I don't see how we could generate this list by introspecting the `run` method.  `inspect.getargspec` just tells us that it takes kwargs.  I guess we could annotate the docstring in some way and then parse it with a package like `docstring-parser` and look for some kind of remote keyword on each documented argument, but that seems pretty gross, and no more maintainable than having this whitelist.

I moved the list to a variable outside the function so that at least if some future code needs to know its content, it can read it.  Please let me know if you've got a better idea how to do this.

Reply:
> parse it with a package like docstring-parser

![](http://img3.wikia.nocookie.net/__cb20150424014315/degrassi/images/5/5a/Cover_eyes_shining.gif)

Reply:
> I moved the list to a variable outside the function so that at least if some future code needs to know its content, it can read it

I don't have any better ideas off the top of my head (other than runner itself actually itemizing its argspec instead of taking in arbitrary `**kwargs`).  Moving this into a defined list is fine for now, I was just curious where this list came from.